src/disc/stk/Albany_GenericSTKFieldContainer.cpp

```diff
@@ -91,107 +91,84 @@ addStateStructs(const Teuchos::RCP<StateInfoStruct>& sis)
     return;
 
   // Code to parse the vector of StateStructs and create STK fields
-  for(std::size_t i = 0; i < sis->size(); i++) {
-    StateStruct& st = *((*sis)[i]);
-    StateStruct::FieldDims& dim = st.dim;
+  for (const auto& st : *sis) {
 
-    if(st.entity == StateStruct::ElemData){
-
-      if (dim.size()==1)
-      {
+    StateStruct::FieldDims& dim = st->dim;
+    if (st->entity == StateStruct::ElemData) {
+      if (dim.size()==1) {
         // Scalar on cell
-        cell_scalar_states.push_back(& metaData->declare_field< double >(stk::topology::ELEMENT_RANK, st.name));
+        cell_scalar_states.push_back(& metaData->declare_field< double >(stk::topology::ELEMENT_RANK, st->name));
         stk::mesh::put_field_on_mesh(*cell_scalar_states.back(), metaData->universal_part(), 1, nullptr);
-        set_output_role(*cell_scalar_states.back(),st.output);
+        set_output_role(*cell_scalar_states.back(),st->output);
       } else if (dim.size()==2) {
         // Vector on cell
-        cell_vector_states.push_back(& metaData->declare_field< double >(stk::topology::ELEMENT_RANK, st.name));
+        cell_vector_states.push_back(& metaData->declare_field< double >(stk::topology::ELEMENT_RANK, st->name));
         stk::mesh::put_field_on_mesh(*cell_vector_states.back(), metaData->universal_part(), dim[1], nullptr);
-        set_output_role(*cell_vector_states.back(),st.output);
+        set_output_role(*cell_vector_states.back(),st->output);
       } else if (dim.size()==3) {
         // 2nd order tensor on cell
-        cell_tensor_states.push_back(& metaData->declare_field< double >(stk::topology::ELEMENT_RANK, st.name));
+        cell_tensor_states.push_back(& metaData->declare_field< double >(stk::topology::ELEMENT_RANK, st->name));
         stk::mesh::put_field_on_mesh(*cell_tensor_states.back(), metaData->universal_part(), dim[2], dim[1], nullptr);
-        set_output_role(*cell_tensor_states.back(),st.output);
+        set_output_role(*cell_tensor_states.back(),st->output);
       } else {
-        TEUCHOS_TEST_FOR_EXCEPTION (true, std::logic_error, "Error! Unexpected state rank.\n");
+        throw std::logic_error("Error! Unexpected state rank.\n");
```

Comment:
why are you replacin `TEUCHOS_TEST_FOR_EXCEPTION` with `throw`?

Reply:
I just find it silly to use a macro to just throw an exception. The macro is meant to wrap a condition check, which should throw the given exception type if the check fails. Hard-coding `false` is just a convoluted way to throw the exception. At this point, why not just throwing the exception by hand?

Reply:
If you dislike this, I can roll back the change.

Reply:
Sounds reasonable. I guess one difference is that throw will throw a message for each process, right?

Reply:
does TEUCHOS_TEST_FOR_EXCEPTION only throw on rank0?

Reply:
Uhm.. maybe not.. I got confused. Here's is what the macro does:
```
{ \
  const bool throw_exception = (throw_exception_test); \
  if(throw_exception) { \
    Teuchos::TestForException_incrThrowNumber(); \
    const int throwNumber = Teuchos::TestForException_getThrowNumber(); \
    std::ostringstream omsg; \
    omsg \
      << __FILE__ << ":" << __LINE__ << ":\n\n" \
      << "Throw number = " << throwNumber \
      << "\n\n" \
      << "Throw test that evaluated to true: "#throw_exception_test \
      << "\n\n" \
      << msg; \
    const std::string &omsgstr = omsg.str(); \
    TEUCHOS_STORE_STACKTRACE(); \
    Teuchos::TestForException_break(omsgstr, throwNumber); \
    throw Exception(omsgstr); \
  } \
}
```

Reply:
Yeah, I _think_ Teuchos simply wraps the message formatting and gives a point where one can break wtih gdb.

I'm going to merge as is, I can revert in the next PR I submit.